### PR TITLE
Reporting Parameter Update for IA And Quality Sections

### PR DESCRIPTION
**Pull request by saquino0827**
### Information
- QPPCT-863

### Changes proposed in this PR:
- Updates the Reporting parameters for IA and Quality Measure sections as such: 
    - Reporting Parameters are now reported in IA Measures and Quality Measure Reference and Results OR their respective sections.
        - If specified on the Measure level, a performance period must be specified for each measure.
        - If specified on the Section level, only one performance period is allowed.
        - If specified in Measure and Section level, a validation error occurs.
- Encoding for Reporting parameters has been moved to be in each IA measure and each Quality measure reference and result measure.
- Tests have been updated and new tests have been added for the new scenarios.
- Updates Quality Measure Sections to require a minimum of 1 Quality Measure Reference and Results. I discovered this as a requirement in the IG: https://ecqi.healthit.gov/system/files/eCQM_QRDA_EC-508_0.pdf#page=33

### Checklist 
- [x] All JUnit tests pass (`mvn clean verify`).
- [x] New unit tests written to cover new functionality.
- [x] Added and updated JavaDocs for non-test classes and methods.
- [x] No local design debt. Do you feel that something is "ugly" after your changes?
- [x] Updated documentation (`README.md`, etc.) depending if the changes require it.

### CODE REVIEW

File: commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java [hunk 1]
```diff
@@ -155,7 +155,27 @@
 	MEASURES_RNR_WITH_DUPLICATED_MEASURE_GUID(71, "Two or more different measure section measure reference and results have "
 		+ "the same measure GUID. Please ensure that each measure section measure reference and results do not have "
 		+ "the same measure GUID."),
-	PERFORMANCE_RATE_MISSING(72, "The Performance Rate is missing");
+	PERFORMANCE_RATE_MISSING(72, "The Performance Rate is missing"),
+	IA_SECTION_IA_MEASURE_DUPLICATE_REPORTING_PARAM(73, "A Reporting Parameter Act has been found in both the "
+		+ "IA section and in IA section measures. The IA Section must have one Reporting Parameters Act "
```
Comment (omnivion) [inline commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java]: should this say IA section and IA measures

File: converter/src/main/java/gov/cms/qpp/conversion/encode/AciSectionEncoder.java [hunk 2]
```diff
@@ -47,7 +47,9 @@ public void internalEncode(JsonWrapper wrapper, Node node) {
 		wrapper.putObject("measurements", measurementsWrapper);
 
 		Optional.ofNullable(node.getParent()).ifPresent(parent -> pilferParent(wrapper, parent));
-		encodeReportingParameter(wrapper, node);
+		if (node.getType().equals(TemplateId.ACI_SECTION)) {
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/encode/AciSectionEncoder.java]: ==

File: converter/src/main/java/gov/cms/qpp/conversion/validate/IaSectionValidator.java [hunk 3]
```diff
@@ -20,7 +20,17 @@
 	protected void internalValidateSingleNode(Node node) {
 		check(node)
 				.childMinimum(ErrorCode.IA_SECTION_MISSING_IA_MEASURE, 1, TemplateId.IA_MEASURE)
-				.childExact(ErrorCode.IA_SECTION_MISSING_REPORTING_PARAM, 1, TemplateId.REPORTING_PARAMETERS_ACT)
-				.onlyHasChildren(ErrorCode.IA_SECTION_WRONG_CHILD, TemplateId.IA_MEASURE, TemplateId.REPORTING_PARAMETERS_ACT);
+				.onlyHasChildren(ErrorCode.IA_SECTION_WRONG_CHILD, TemplateId.IA_MEASURE, TemplateId.REPORTING_PARAMETERS_ACT)
+			    .noParentChildDuplications(ErrorCode.IA_SECTION_IA_MEASURE_DUPLICATE_REPORTING_PARAM,
+				    TemplateId.REPORTING_PARAMETERS_ACT, TemplateId.IA_MEASURE);
+
+		if (node.getChildNodes(TemplateId.REPORTING_PARAMETERS_ACT).count() > 0) {
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/IaSectionValidator.java]: Using findAny().isPresent() could be better here -- count() makes the stream do the full computation

File: commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java (hunk 1 above)
Comment (saquino0827) [reply commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java]: Yes. Ultimately, the Reporting Parameter Act can be specified once within the IA Section or specified in each individual IA measure. However, the spec does not allow for the Reporting Parameter Act to be specified in both the IA section and within the IA measures of the given IA section.

File: converter/src/main/java/gov/cms/qpp/conversion/validate/IaSectionValidator.java (hunk 3 above)
Comment (saquino0827) [reply converter/src/main/java/gov/cms/qpp/conversion/validate/IaSectionValidator.java]: 👍  will change.

File: converter/src/main/java/gov/cms/qpp/conversion/encode/IaMeasureEncoder.java [hunk 4]
```diff
@@ -1,6 +1,7 @@
 package gov.cms.qpp.conversion.encode;
 
 import gov.cms.qpp.conversion.Context;
+import gov.cms.qpp.conversion.decode.ReportingParametersActDecoder;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/encode/IaMeasureEncoder.java]: MINOR Remove this unused import 'gov.cms.qpp.conversion.decode.ReportingParametersActDecoder'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: converter/src/main/java/gov/cms/qpp/conversion/encode/QualityMeasureIdEncoder.java [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import gov.cms.qpp.conversion.Context;
 import gov.cms.qpp.conversion.decode.AggregateCountDecoder;
 import gov.cms.qpp.conversion.decode.MeasureDataDecoder;
+import gov.cms.qpp.conversion.decode.ReportingParametersActDecoder;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/encode/QualityMeasureIdEncoder.java]: MINOR Remove this unused import 'gov.cms.qpp.conversion.decode.ReportingParametersActDecoder'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java [hunk 6]
```diff
@@ -7,6 +7,7 @@
 import gov.cms.qpp.conversion.model.Node;
 import gov.cms.qpp.conversion.model.TemplateId;
 import gov.cms.qpp.conversion.model.error.Detail;
+import gov.cms.qpp.conversion.model.error.ErrorCode;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java]: MINOR Remove this unused import 'gov.cms.qpp.conversion.model.error.ErrorCode'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: converter/src/main/java/gov/cms/qpp/conversion/validate/QualityMeasureSectionValidator.java [hunk 7]
```diff
@@ -4,6 +4,7 @@
 import gov.cms.qpp.conversion.model.Node;
 import gov.cms.qpp.conversion.model.TemplateId;
 import gov.cms.qpp.conversion.model.Validator;
+import gov.cms.qpp.conversion.model.error.Error;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/QualityMeasureSectionValidator.java]: MINOR Remove this unused import 'gov.cms.qpp.conversion.model.error.Error'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by FlexionSonarQube** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ErrorCode.java#L79](https://github.com/CMSgov/qpp-conversion-tool/blob/f30b883ba21781dbff690366893b0ee8b3582133/commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java#L79): Define a constant instead of duplicating this literal "Please ensure the Reporting Parameters Act complies with the Implementation Guide (IG). " 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ErrorCode.java#L80](https://github.com/CMSgov/qpp-conversion-tool/blob/f30b883ba21781dbff690366893b0ee8b3582133/commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java#L80): Define a constant instead of duplicating this literal "Here is a link to the IG Reporting Parameter Act section: " 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by FlexionSonarQube** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ErrorCode.java#L79](https://github.com/CMSgov/qpp-conversion-tool/blob/1cba1b42cb1d6c384ad9aabd5c62528b000a72f5/commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java#L79): Define a constant instead of duplicating this literal "Please ensure the Reporting Parameters Act complies with the Implementation Guide (IG). " 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ErrorCode.java#L80](https://github.com/CMSgov/qpp-conversion-tool/blob/1cba1b42cb1d6c384ad9aabd5c62528b000a72f5/commons/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java#L80): Define a constant instead of duplicating this literal "Here is a link to the IG Reporting Parameter Act section: " 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**